feat: slash proposers properly on failed epoch

just-mitch opened the pull request:
This PR enhances the slashing mechanism for epoch pruning by distinguishing between two different offenses:

1. `DATA_WITHHOLDING` - When transactions for a pruned epoch are not available
2. `VALID_EPOCH_PRUNED` - When transactions are available and the epoch could have been proven

The implementation:
- Adds transaction collection and block validation to the `EpochPruneWatcher`
- Updates the `L2PruneDetected` event to include pruned blocks instead of just block numbers
- Exposes the `TxCollector` interface to allow reuse outside of P2P
- Refactors the `BlockBuilder` to support external fork management
- Moves validator errors to the stdlib for broader reuse

Fix #14407 
Fix #14408

just-mitch committed:
feat: slash proposers properly on failed epoch

LHerskind commented on yarn-project/slasher/src/epoch_prune_watcher.test.ts:
```diff
@@ -92,16 +112,113 @@ describe('EpochPruneWatcher', () => {
+      type: L2BlockSourceEvents.L2PruneDetected,
+    });
+
+    await sleep(0);
```
?

LHerskind commented on yarn-project/slasher/src/epoch_prune_watcher.test.ts:
```diff
@@ -92,16 +112,113 @@ describe('EpochPruneWatcher', () => {
+      type: L2BlockSourceEvents.L2PruneDetected,
+    });
+
+    await sleep(0);
```
?

LHerskind commented on yarn-project/end-to-end/src/e2e_p2p/shared.ts:
```diff
@@ -90,3 +105,123 @@ export async function createPXEServiceAndPrepareTransactions(
+  logger.info(`Waiting for slash proposal to be executed`);
+  await awaitProposalExecution(slashingProposer, slashingRoundSize * 2 * aztecSlotDuration);
+
+  // The attesters should still form the committee
```
This is fairly brittle I would think. The only reason that it is still working for you here would likely be that we are drawing the committee based on lagging data, so if the duration to execute the proposal was larger etc you might just fail here.

just-mitch replied:
I'm not sure I see the problem- the above `awaitProposal` will wait *up to* 2 rounds, but returns as soon as the thing is executed, so `getCurrentEpochCommittee` should always return the committee with the slashed attesters.

LHerskind reviewed (approved):
Think most seems fine, have a few comments mainly around how it will break with #14833

Maddiaa0 commented on yarn-project/stdlib/src/block/l2_block_source.ts:
```diff
@@ -179,9 +179,15 @@ export enum L2BlockSourceEvents {
+export type L2BlockPruneEvent = {
+  type: 'l2PruneDetected';
+  epochNumber: bigint;
+  blocks: L2Block[];
```
it should just pass pointers around for the object types

just-mitch commented on yarn-project/archiver/src/archiver/archiver.ts:
```diff
@@ -391,15 +391,17 @@ export class Archiver extends (EventEmitter as new () => ArchiverEmitter) implem
-        blockNumber: pruneFrom,
-        slotNumber: pruneFromSlotNumber,
         epochNumber: pruneFromEpochNumber,
+        blocks,
```
Yes, I thought about that. But in reality if we just assume that a transaction effect contains 100 `Fr` so 3.2KB, (should be super conservative) and then at 1 TPS a block is 36 so 115KB, so an entire epoch is less than 4MB. So should be safe to sustain a temporary increase of 4MB here. I suspect the real increase in RAM will come during the replay of the epoch as the fork we create grows.

But the important thing is that we *need* to make a copy of the data here, as we are about to prune it all in `this.store.unwindBlocks`.

just-mitch committed:
review comments and fix minor conflict in tx collector

just-mitch committed:
slash the deposit amount

just-mitch committed:
chore: stop voting for override after executed (#14880)

This PR fixes an issue where the SlasherClient would continue to signal
an override payload even after it had been executed. Now, when a
proposal is executed, the client checks if it matches the override
payload and disables the override if it does.

Changes include:
- Added an `overridePayloadActive` flag to track whether the override
payload should be used
- Modified `proposalExecuted()` to disable the override when the
override payload is executed
- Made `proposalExecuted()` protected instead of private to allow
testing
- Added tests to verify the override payload behavior
- Created a `TestSlasherClient` class to expose protected methods for
testing

Fix #14488

just-mitch committed:
fix flake in data withholding test

just-mitch committed:
fix flake
